codex/sales/states/proving.nim

```diff
@@ -62,24 +62,31 @@ proc proveLoop(
 
   proc waitUntilPeriod(period: Period) {.async.} =
     let periodicity = await market.periodicity()
-    await clock.waitUntil(periodicity.periodStart(period).truncate(int64))
+    await clock.waitUntil(periodicity.periodStart(period).truncate(int64) + 1)
 
   while true:
     let currentPeriod = await getCurrentPeriod()
     let slotState = await market.slotState(slot.id)
-    if slotState == SlotState.Finished:
-      debug "Slot reached finished state", period = currentPeriod
-      return
-
-    if slotState != SlotState.Filled:
-      raise newException(SlotNotFilledError, "Slot is not in Filled state!")
 
-    debug "Proving for new period", period = currentPeriod
-
-    if (await market.isProofRequired(slotId)) or (await market.willProofBeRequired(slotId)):
-      let challenge = await market.getChallenge(slotId)
-      debug "Proof is required", period = currentPeriod, challenge = challenge
-      await state.prove(slot, challenge, onProve, market, currentPeriod)
+    case slotState
+    of SlotState.Filled:
+      debug "Proving for new period", period = currentPeriod
+      if (await market.isProofRequired(slotId)) or (await market.willProofBeRequired(slotId)):
+        let challenge = await market.getChallenge(slotId)
+        debug "Proof is required", period = currentPeriod, challenge = challenge
+        await state.prove(slot, challenge, onProve, market, currentPeriod)
+    of SlotState.Cancelled:
+      debug "Slot reached cancelled state"
+      # do nothing, let onCancelled callback take care of it
+    of SlotState.Failed:
```

Comment:
Does proving-loop get cleaned up correctly in case we leave this state via Cancelled or Failed callbacks?